// weboptimizer/browserAPI.compiled.js
// #!/usr/bin/env node

// -*- coding: utf-8 -*-
'use strict';
/* !
    region header
    Copyright Torben Sickert (info["~at~"]torben.website) 16.12.2012

    License
    -------

    This library written by Torben Sickert stand under a creative commons naming
    3.0 unported license. see http://creativecommons.org/licenses/by/3.0/deed.de
    endregion
*/
// region imports
/* eslint-disable no-unused-vars */

Object.defineProperty(exports, "__esModule", {
    value: true
});

// endregion
// region variables
var onCreatedListener = [];
/* eslint-enable no-unused-vars */
// endregion
// region declaration

var browserAPI = void 0;
// endregion
// region ensure presence of common browser environment
if (typeof TARGET_TECHNOLOGY === 'undefined' || TARGET_TECHNOLOGY === 'node') {
    (function () {
        // region mock browser environment
        var path = require('path');
        var metaDOM = require('jsdom');
        var virtualConsole = metaDOM.createVirtualConsole().sendTo(console, { omitJsdomErrors: true });
        virtualConsole.on('jsdomError', function (error) {
            if (!browserAPI.debug && ['XMLHttpRequest', 'resource loading'
            // IgnoreTypeCheck
            ].includes(error.type)) console.warn('Loading resource failed: ' + error.toString() + '.');else
                // IgnoreTypeCheck
                console.error(error.stack, error.detail);
        });
        var template = void 0;
        if (typeof NAME === 'undefined' || NAME === 'webOptimizer') template = require('pug').compileFile(path.join(__dirname, 'index.pug'), { pretty: true })({ configuration: {
                name: 'test', givenCommandLineArguments: []
            } });else
            // IgnoreTypeCheck
            template = require('webOptimizerDefaultTemplateFilePath');
        metaDOM.env({
            created: function created(error, window) {
                browserAPI = {
                    debug: false, domContentLoaded: false, metaDOM: metaDOM, window: window,
                    windowLoaded: false };
                browserAPI.window.document.addEventListener('DOMContentLoaded', function () {
                    browserAPI.domContentLoaded = true;
                });
                if (error) throw error;else {
                    var _iteratorNormalCompletion = true;
                    var _didIteratorError = false;
                    var _iteratorError = undefined;

                    try {
                        for (var _iterator = onCreatedListener[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                            var _callback = _step.value;

                            _callback(browserAPI, false);
                        }
                    } catch (err) {
                        _didIteratorError = true;
                        _iteratorError = err;
                    } finally {
                        try {
                            if (!_iteratorNormalCompletion && _iterator.return) {
                                _iterator.return();
                            }
                        } finally {
                            if (_didIteratorError) {
                                throw _iteratorError;
                            }
                        }
                    }
                }
            },
            features: {
                FetchExternalResources: ['script', 'frame', 'iframe', 'link', 'img'],
                ProcessExternalResources: ['script'],
                SkipExternalResources: false
            },
            html: template,
            onload: function onload() {
                browserAPI.domContentLoaded = true;
                browserAPI.windowLoaded = true;
            },
            resourceLoader: function resourceLoader(resource, callback) {
                if (resource.url.hostname === 'localhost') {
                    resource.url.host = resource.url.hostname = '';
                    resource.url.port = null;
                    resource.url.protocol = 'file:';
                    resource.url.href = resource.url.href.replace(/^[a-zA-Z]+:\/\/localhost(?::[0-9]+)?/, 'file://' + process.cwd());
                    resource.url.path = resource.url.pathname = path.join(process.cwd(), resource.url.path);
                }
                if (browserAPI.debug) console.info('Load resource "' + resource.url.href + '".');
                return resource.defaultFetch(function (error) {
                    if (!error) callback.apply(this, arguments);
                });
            },
            url: 'http://localhost',
            virtualConsole: virtualConsole
        });
        // endregion
    })();
} else {
    browserAPI = {
        debug: false, domContentLoaded: false, metaDOM: null, window: window,
        windowLoaded: false };
    window.document.addEventListener('DOMContentLoaded', function () {
        browserAPI.domContentLoaded = true;
        var _iteratorNormalCompletion2 = true;
        var _didIteratorError2 = false;
        var _iteratorError2 = undefined;

        try {
            for (var _iterator2 = onCreatedListener[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
                var callback = _step2.value;

                callback(browserAPI, false);
            }
        } catch (err) {
            _didIteratorError2 = true;
            _iteratorError2 = err;
        } finally {
            try {
                if (!_iteratorNormalCompletion2 && _iterator2.return) {
                    _iterator2.return();
                }
            } finally {
                if (_didIteratorError2) {
                    throw _iteratorError2;
                }
            }
        }
    });
    window.addEventListener('load', function () {
        browserAPI.windowLoaded = true;
    });
}
// endregion

exports.default = function (callback) {
    var clear = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];

    // region initialize global context
    /*
        NOTE: We have to define window globally before anything is loaded to
        ensure that all future instances share the same window object.
    */
    if (clear && typeof global !== 'undefined' && global !== browserAPI.window) global.window = browserAPI.window;
    // endregion
    if (typeof TARGET_TECHNOLOGY === 'undefined' || TARGET_TECHNOLOGY === 'node') return browserAPI ? callback(browserAPI, true) : onCreatedListener.push(callback);
    return browserAPI.domContentLoaded ? callback(browserAPI, true) : onCreatedListener.push(callback);
};
// region vim modline
// vim: set tabstop=4 shiftwidth=4 expandtab:
// vim: foldmethod=marker foldmarker=region,endregion:
// endregion
