// weboptimizer/helper.compiled.js
#!/usr/bin/env node


// -*- coding: utf-8 -*-
'use strict';
/* !
    region header
    Copyright Torben Sickert (info["~at~"]torben.website) 16.12.2012

    License
    -------

    This library written by Torben Sickert stand under a creative commons naming
    3.0 unported license. see http://creativecommons.org/licenses/by/3.0/deed.de
    endregion
*/
// region imports

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _child_process = require('child_process');

var _clientnode = require('clientnode');

var _clientnode2 = _interopRequireDefault(_clientnode);

var _fs = require('fs');

var fileSystem = _interopRequireWildcard(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

// NOTE: Only needed for debugging this file.
try {
    require('source-map-support/register');
} catch (error) {}

// endregion
// region methods
/**
 * Provides a class of static methods with generic use cases.
 */
var Helper = function () {
    function Helper() {
        _classCallCheck(this, Helper);
    }

    _createClass(Helper, null, [{
        key: 'isFilePathInLocation',

        // region boolean
        /**
         * Determines whether given file path is within given list of file
         * locations.
         * @param filePath - Path to file to check.
         * @param locationsToCheck - Locations to take into account.
         * @returns Value "true" if given file path is within one of given
         * locations or "false" otherwise.
         */
        value: function isFilePathInLocation(filePath, locationsToCheck) {
            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
                for (var _iterator = locationsToCheck[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                    var pathToCheck = _step.value;

                    if (_path2.default.resolve(filePath).startsWith(_path2.default.resolve(pathToCheck))) return true;
                }
            } catch (err) {
                _didIteratorError = true;
                _iteratorError = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion && _iterator.return) {
                        _iterator.return();
                    }
                } finally {
                    if (_didIteratorError) {
                        throw _iteratorError;
                    }
                }
            }

            return false;
        }
        // endregion
        // region string
        /**
         * Strips loader informations form given module request including loader
         * prefix and query parameter.
         * @param moduleID - Module request to strip.
         * @returns Given module id stripped.
         */

    }, {
        key: 'stripLoader',
        value: function stripLoader(moduleID) {
            moduleID = moduleID.toString();
            var moduleIDWithoutLoader = moduleID.substring(moduleID.lastIndexOf('!') + 1);
            return moduleIDWithoutLoader.includes('?') ? moduleIDWithoutLoader.substring(0, moduleIDWithoutLoader.indexOf('?')) : moduleIDWithoutLoader;
        }
        // endregion
        // region array
        /**
         * Converts given list of path to a normalized list with unique values.
         * @param paths - File paths.
         * @returns The given file path list with normalized unique values.
         */

    }, {
        key: 'normalizePaths',
        value: function normalizePaths(paths) {
            return Array.from(new Set(paths.map(function (givenPath) {
                givenPath = _path2.default.normalize(givenPath);
                if (givenPath.endsWith('/')) return givenPath.substring(0, givenPath.length - 1);
                return givenPath;
            })));
        }
        // endregion
        // region data
        /**
         * Converts given serialized or base64 encoded string into a javaScript
         * one if possible.
         * @param serializedObject - Object as string.
         * @param scope - An optional scope which will be used to evaluate given
         * object in.
         * @param name - The name under given scope will be available.
         * @returns The parsed object if possible and null otherwise.
         */

    }, {
        key: 'parseEncodedObject',
        value: function parseEncodedObject(serializedObject) {
            var scope = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];
            var name = arguments.length <= 2 || arguments[2] === undefined ? 'scope' : arguments[2];

            if (!serializedObject.startsWith('{')) serializedObject = Buffer.from(serializedObject, 'base64').toString('utf8');
            try {
                // IgnoreTypeCheck
                return new Function(name, 'return ' + serializedObject)(scope);
            } catch (error) {}
            return null;
        }
        // endregion
        // region process handler
        /**
         * Generates a one shot close handler which triggers given promise methods.
         * If a reason is provided it will be given as resolve target. An Error
         * will be generated if return code is not zero. The generated Error has
         * a property "returnCode" which provides corresponding process return
         * code.
         * @param resolve - Promise's resolve function.
         * @param reject - Promise's reject function.
         * @param reason - Promise target if process has a zero return code.
         * @param callback - Optional function to call of process has successfully
         * finished.
         * @returns Process close handler function.
         */

    }, {
        key: 'getProcessCloseHandler',
        value: function getProcessCloseHandler(resolve, reject) {
            var reason = arguments.length <= 2 || arguments[2] === undefined ? null : arguments[2];
            var callback = arguments.length <= 3 || arguments[3] === undefined ? function () {} : arguments[3];

            var finished = false;
            return function (returnCode) {
                if (!finished) if (typeof returnCode !== 'number' || returnCode === 0) {
                    callback();
                    resolve(reason);
                } else {
                    var error = new Error('Task exited with error code ' + returnCode);
                    // IgnoreTypeCheck
                    error.returnCode = returnCode;
                    reject(error);
                }
                finished = true;
            };
        }
        /**
         * Forwards given child process communication channels to corresponding
         * current process communication channels.
         * @param childProcess - Child process meta data.
         * @returns Given child process meta data.
         */

    }, {
        key: 'handleChildProcess',
        value: function handleChildProcess(childProcess) {
            childProcess.stdout.pipe(process.stdout);
            childProcess.stderr.pipe(process.stderr);
            childProcess.on('close', function (returnCode) {
                if (returnCode !== 0) console.error('Task exited with error code ' + returnCode);
            });
            return childProcess;
        }
        // endregion
        // region file handler
        /**
         * Applies file path/name placeholder replacements with given bundle
         * associated informations.
         * @param filePathTemplate - File path to process placeholder in.
         * @param informations - Scope to use for processing.
         * @returns Processed file path.
         */

    }, {
        key: 'renderFilePathTemplate',
        value: function renderFilePathTemplate(filePathTemplate) {
            var informations = arguments.length <= 1 || arguments[1] === undefined ? {
                '[name]': '.__dummy__', '[id]': '.__dummy__',
                '[hash]': '.__dummy__'
            } : arguments[1];

            var filePath = filePathTemplate;
            for (var placeholderName in informations) {
                if (informations.hasOwnProperty(placeholderName)) filePath = filePath.replace(new RegExp(_clientnode2.default.stringConvertToValidRegularExpression(placeholderName), 'g'), informations[placeholderName]);
            }return filePath;
        }
        /**
         * Check if given request points to an external dependency not maintained
         * by current package context.
         * @param request - Request to determine.
         * @param context - Context of current project.
         * @param requestContext - Context of given request to resolve relative to.
         * @param normalizedInternalInjection - Mapping of chunk names to modules
         * which should be injected.
         * @param externalModuleLocations - Array if paths where external modules
         * take place.
         * @param moduleAliases - Mapping of aliases to take into account.
         * @param knownExtensions - List of file extensions to take into account.
         * @param referencePath - Path to resolve local modules relative to.
         * @param pathsToIgnore - Paths which marks location to ignore.
         * @param includePattern - Array of regular expressions to explicitly mark
         * as external dependency.
         * @param excludePattern - Array of regular expressions to explicitly mark
         * as internal dependency.
         * @param inPlaceNormalLibrary - Indicates whether normal libraries should
         * be external or not.
         * @param inPlaceDynamicLibrary - Indicates whether requests with
         * integrated loader configurations should be marked as external or not.
         * @param externalHandableFileExtensions - File extensions which should be
         * able to be handled by the external module bundler. If array is empty
         * every extension will be assumed to be supported.
         * @returns A new resolved request indicating whether given request is an
         * external one.
         */

    }, {
        key: 'determineExternalRequest',
        value: function determineExternalRequest(request) {
            var context = arguments.length <= 1 || arguments[1] === undefined ? './' : arguments[1];
            var requestContext = arguments.length <= 2 || arguments[2] === undefined ? './' : arguments[2];
            var normalizedInternalInjection = arguments.length <= 3 || arguments[3] === undefined ? {} : arguments[3];
            var externalModuleLocations = arguments.length <= 4 || arguments[4] === undefined ? [_path2.default.resolve(__dirname, 'node_modules')] : arguments[4];
            var moduleAliases = arguments.length <= 5 || arguments[5] === undefined ? {} : arguments[5];
            var knownExtensions = arguments.length <= 6 || arguments[6] === undefined ? ['', '.js', '.css', '.svg', '.html', 'json'] : arguments[6];
            var referencePath = arguments.length <= 7 || arguments[7] === undefined ? './' : arguments[7];
            var pathsToIgnore = arguments.length <= 8 || arguments[8] === undefined ? ['.git'] : arguments[8];
            var includePattern = arguments.length <= 9 || arguments[9] === undefined ? [] : arguments[9];
            var excludePattern = arguments.length <= 10 || arguments[10] === undefined ? [] : arguments[10];
            var inPlaceNormalLibrary = arguments.length <= 11 || arguments[11] === undefined ? false : arguments[11];
            var inPlaceDynamicLibrary = arguments.length <= 12 || arguments[12] === undefined ? true : arguments[12];
            var externalHandableFileExtensions = arguments.length <= 13 || arguments[13] === undefined ? ['', '.js', '.node', '.json'] : arguments[13];

            context = _path2.default.resolve(context);
            requestContext = _path2.default.resolve(requestContext);
            referencePath = _path2.default.resolve(referencePath);
            // NOTE: We apply alias on externals additionally.
            var resolvedRequest = Helper.applyAliases(request.substring(request.lastIndexOf('!') + 1), moduleAliases);
            /*
                NOTE: Aliases doesn't have to be forwarded since we pass an already
                resolved request.
            */
            var filePath = Helper.determineModuleFilePath(resolvedRequest, {}, knownExtensions, requestContext, referencePath, pathsToIgnore);
            if (!(filePath || inPlaceNormalLibrary) || _clientnode2.default.isAnyMatching(resolvedRequest, includePattern)) return resolvedRequest;
            if (_clientnode2.default.isAnyMatching(resolvedRequest, excludePattern)) return null;
            for (var chunkName in normalizedInternalInjection) {
                if (normalizedInternalInjection.hasOwnProperty(chunkName)) {
                    var _iteratorNormalCompletion2 = true;
                    var _didIteratorError2 = false;
                    var _iteratorError2 = undefined;

                    try {
                        for (var _iterator2 = normalizedInternalInjection[chunkName][Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
                            var _moduleID = _step2.value;

                            if (Helper.determineModuleFilePath(_moduleID, moduleAliases, knownExtensions, context, referencePath, pathsToIgnore) === filePath) return null;
                        }
                    } catch (err) {
                        _didIteratorError2 = true;
                        _iteratorError2 = err;
                    } finally {
                        try {
                            if (!_iteratorNormalCompletion2 && _iterator2.return) {
                                _iterator2.return();
                            }
                        } finally {
                            if (_didIteratorError2) {
                                throw _iteratorError2;
                            }
                        }
                    }
                }
            } /*
                  NOTE: We mark dependencies as external if they does not contain a
                  loader in their request and aren't part of the current main package
                  or have a file extension other than javaScript aware.
              */
            if (!inPlaceNormalLibrary && (externalHandableFileExtensions.length === 0 || filePath && externalHandableFileExtensions.includes(_path2.default.extname(filePath)) || !filePath && externalHandableFileExtensions.includes('')) && !(inPlaceDynamicLibrary && request.includes('!')) && (!filePath && inPlaceDynamicLibrary || filePath && (!filePath.startsWith(context) || Helper.isFilePathInLocation(filePath, externalModuleLocations)))) return resolvedRequest;
            return null;
        }
        /**
         * Checks if given path points to a valid file.
         * @param filePath - Path to file.
         * @returns A boolean which indicates file existents.
         */

    }, {
        key: 'isFileSync',
        value: function isFileSync(filePath) {
            try {
                fileSystem.accessSync(filePath, fileSystem.F_OK);
                return true;
            } catch (error) {
                return false;
            }
        }
        /**
         * Iterates through given directory structure recursively and calls given
         * callback for each found file. Callback gets file path and corresponding
         * stat object as argument.
         * @param directoryPath - Path to directory structure to traverse.
         * @param callback - Function to invoke for each traversed file.
         * @returns Given callback function.
         */

    }, {
        key: 'walkDirectoryRecursivelySync',
        value: function walkDirectoryRecursivelySync(directoryPath) {
            var callback = arguments.length <= 1 || arguments[1] === undefined ? function (_filePath, _stat) {
                return true;
            } : arguments[1];

            fileSystem.readdirSync(directoryPath).forEach(function (fileName) {
                var filePath = _path2.default.resolve(directoryPath, fileName);
                var stat = fileSystem.statSync(filePath);
                if (callback(filePath, stat) !== false && stat && stat.isDirectory()) Helper.walkDirectoryRecursivelySync(filePath, callback);
            });
            return callback;
        }
        /**
         * Copies given source file via path to given target directory location
         * with same target name as source file has or copy to given complete
         * target file path.
         * @param sourcePath - Path to file to copy.
         * @param targetPath - Target directory or complete file location to copy
         * to.
         * @returns Determined target file path.
         */

    }, {
        key: 'copyFileSync',
        value: function copyFileSync(sourcePath, targetPath) {
            /*
                NOTE: If target path references a directory a new file with the
                same name will be created.
            */
            try {
                if (fileSystem.lstatSync(targetPath).isDirectory()) targetPath = _path2.default.resolve(targetPath, _path2.default.basename(sourcePath));
            } catch (error) {}
            fileSystem.writeFileSync(targetPath, fileSystem.readFileSync(sourcePath));
            return targetPath;
        }
        /**
         * Copies given source directory via path to given target directory
         * location with same target name as source file has or copy to given
         * complete target directory path.
         * @param sourcePath - Path to directory to copy.
         * @param targetPath - Target directory or complete directory location to
         * copy in.
         * @returns Determined target directory path.
         */

    }, {
        key: 'copyDirectoryRecursiveSync',
        value: function copyDirectoryRecursiveSync(sourcePath, targetPath) {
            try {
                // Check if folder needs to be created or integrated.
                if (fileSystem.lstatSync(targetPath).isDirectory()) targetPath = _path2.default.resolve(targetPath, _path2.default.basename(sourcePath));
            } catch (error) {}
            fileSystem.mkdirSync(targetPath);
            Helper.walkDirectoryRecursivelySync(sourcePath, function (currentSourcePath, stat) {
                var currentTargetPath = _path2.default.join(targetPath, currentSourcePath.substring(sourcePath.length));
                if (stat.isDirectory()) fileSystem.mkdirSync(currentTargetPath);else Helper.copyFileSync(currentSourcePath, currentTargetPath);
            });
            return targetPath;
        }
        /**
         * Determines a asset type if given file.
         * @param filePath - Path to file to analyse.
         * @param buildConfiguration - Meta informations for available asset
         * types.
         * @param paths - List of paths to search if given path doesn't reference
         * a file directly.
         * @returns Determined file type or "null" of given file couldn't be
         * determined.
         */

    }, {
        key: 'determineAssetType',
        value: function determineAssetType(filePath, buildConfiguration, paths) {
            var result = null;
            for (var type in buildConfiguration) {
                if (_path2.default.extname(filePath) === '.' + buildConfiguration[type].extension) {
                    result = type;
                    break;
                }
            }if (!result) {
                var _arr = ['source', 'target'];

                for (var _i = 0; _i < _arr.length; _i++) {
                    var _type = _arr[_i];
                    for (var assetType in paths[_type].asset) {
                        if (paths[_type].asset.hasOwnProperty(assetType) && assetType !== 'base' && paths[_type].asset[assetType] && filePath.startsWith(paths[_type].asset[assetType])) return assetType;
                    }
                }
            }return result;
        }
        /**
         * Adds a property with a stored array of all matching file paths, which
         * matches each build configuration in given entry path and converts given
         * build configuration into a sorted array were javaScript files takes
         * precedence.
         * @param configuration - Given build configurations.
         * @param entryPath - Path to analyse nested structure.
         * @param pathsToIgnore - Paths which marks location to ignore.
         * @returns Converted build configuration.
         */

    }, {
        key: 'resolveBuildConfigurationFilePaths',
        value: function resolveBuildConfigurationFilePaths(configuration) {
            var entryPath = arguments.length <= 1 || arguments[1] === undefined ? './' : arguments[1];
            var pathsToIgnore = arguments.length <= 2 || arguments[2] === undefined ? ['.git'] : arguments[2];

            var buildConfiguration = [];
            var index = 0;
            for (var type in configuration) {
                if (configuration.hasOwnProperty(type)) {
                    var newItem = _clientnode2.default.extendObject(true, { filePaths: [] }, configuration[type]);
                    Helper.walkDirectoryRecursivelySync(entryPath, function (index, buildConfigurationItem) {
                        return function (filePath, stat) {
                            if (Helper.isFilePathInLocation(filePath, pathsToIgnore)) return false;
                            if (stat.isFile() && _path2.default.extname(filePath).substring(1) === buildConfigurationItem.extension && !new RegExp(buildConfigurationItem.fileNamePattern).test(filePath)) buildConfigurationItem.filePaths.push(filePath);
                        };
                    }(index, newItem));
                    buildConfiguration.push(newItem);
                    index += 1;
                }
            }return buildConfiguration.sort(function (first, second) {
                if (first.outputExtension !== second.outputExtension) {
                    if (first.outputExtension === 'js') return -1;
                    if (second.outputExtension === 'js') return 1;
                    return first.outputExtension < second.outputExtension ? -1 : 1;
                }
                return 0;
            });
        }
        /**
         * Determines all file and directory paths related to given internal
         * modules as array.
         * @param internalInjection - List of module ids or module file paths.
         * @param moduleAliases - Mapping of aliases to take into account.
         * @param knownExtensions - List of file extensions to take into account.
         * @param context - File path to resolve relative to.
         * @param referencePath - Path to search for local modules.
         * @param pathsToIgnore - Paths which marks location to ignore.
         * @param relativeModuleFilePaths - Module file paths relatively to given
         * context.
         * @param packageEntryFileNames - Names of possible package entry files.
         * @returns Object with a file path and directory path key mapping to
         * corresponding list of paths.
         */

    }, {
        key: 'determineModuleLocations',
        value: function determineModuleLocations(internalInjection) {
            var moduleAliases = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];
            var knownExtensions = arguments.length <= 2 || arguments[2] === undefined ? ['', '.js', '.css', '.svg', '.html', 'json'] : arguments[2];
            var context = arguments.length <= 3 || arguments[3] === undefined ? './' : arguments[3];
            var referencePath = arguments.length <= 4 || arguments[4] === undefined ? '' : arguments[4];
            var pathsToIgnore = arguments.length <= 5 || arguments[5] === undefined ? ['.git'] : arguments[5];
            var relativeModuleFilePaths = arguments.length <= 6 || arguments[6] === undefined ? ['', 'node_modules', '../'] : arguments[6];
            var packageEntryFileNames = arguments.length <= 7 || arguments[7] === undefined ? ['__package__', '', 'index', 'main'] : arguments[7];

            var filePaths = [];
            var directoryPaths = [];
            var normalizedInternalInjection = Helper.normalizeInternalInjection(internalInjection);
            for (var chunkName in normalizedInternalInjection) {
                if (normalizedInternalInjection.hasOwnProperty(chunkName)) {
                    var _iteratorNormalCompletion3 = true;
                    var _didIteratorError3 = false;
                    var _iteratorError3 = undefined;

                    try {
                        for (var _iterator3 = normalizedInternalInjection[chunkName][Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
                            var _moduleID2 = _step3.value;

                            var filePath = Helper.determineModuleFilePath(_moduleID2, moduleAliases, knownExtensions, context, referencePath, pathsToIgnore, relativeModuleFilePaths, packageEntryFileNames);
                            if (filePath) {
                                filePaths.push(filePath);
                                var directoryPath = _path2.default.dirname(filePath);
                                if (!directoryPaths.includes(directoryPath)) directoryPaths.push(directoryPath);
                            }
                        }
                    } catch (err) {
                        _didIteratorError3 = true;
                        _iteratorError3 = err;
                    } finally {
                        try {
                            if (!_iteratorNormalCompletion3 && _iterator3.return) {
                                _iterator3.return();
                            }
                        } finally {
                            if (_didIteratorError3) {
                                throw _iteratorError3;
                            }
                        }
                    }
                }
            }return { filePaths: filePaths, directoryPaths: directoryPaths };
        }
        /**
         * Every injection definition type can be represented as plain object
         * (mapping from chunk name to array of module ids). This method converts
         * each representation into the normalized plain object notation.
         * @param internalInjection - Given internal injection to normalize.
         * @returns Normalized representation of given internal injection.
         */

    }, {
        key: 'normalizeInternalInjection',
        value: function normalizeInternalInjection(internalInjection) {
            var result = {};
            if (internalInjection instanceof Object && _clientnode2.default.isPlainObject(internalInjection)) {
                var hasContent = false;
                var chunkNamesToDelete = [];
                for (var chunkName in internalInjection) {
                    if (internalInjection.hasOwnProperty(chunkName)) if (Array.isArray(internalInjection[chunkName])) {
                        if (internalInjection[chunkName].length > 0) {
                            hasContent = true;
                            result[chunkName] = internalInjection[chunkName];
                        } else chunkNamesToDelete.push(chunkName);
                    } else {
                        hasContent = true;
                        result[chunkName] = [internalInjection[chunkName]];
                    }
                }if (hasContent) {
                    var _iteratorNormalCompletion4 = true;
                    var _didIteratorError4 = false;
                    var _iteratorError4 = undefined;

                    try {
                        for (var _iterator4 = chunkNamesToDelete[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
                            var _chunkName = _step4.value;

                            delete result[_chunkName];
                        }
                    } catch (err) {
                        _didIteratorError4 = true;
                        _iteratorError4 = err;
                    } finally {
                        try {
                            if (!_iteratorNormalCompletion4 && _iterator4.return) {
                                _iterator4.return();
                            }
                        } finally {
                            if (_didIteratorError4) {
                                throw _iteratorError4;
                            }
                        }
                    }
                } else result = { index: [] };
            } else if (typeof internalInjection === 'string') result = { index: [internalInjection] };else if (Array.isArray(internalInjection)) result = { index: internalInjection };
            return result;
        }
        /**
         * Determines all concrete file paths for given injection which are marked
         * with the "__auto__" indicator.
         * @param givenInjection - Given internal and external injection to take
         * into account.
         * @param buildConfigurations - Resolved build configuration.
         * @param modulesToExclude - A list of modules to exclude (specified by
         * path or id) or a mapping from chunk names to module ids.
         * @param moduleAliases - Mapping of aliases to take into account.
         * @param knownExtensions - File extensions to take into account.
         * @param context - File path to use as starting point.
         * @param referencePath - Reference path from where local files should be
         * resolved.
         * @param pathsToIgnore - Paths which marks location to ignore.
         * @returns Given injection with resolved marked indicators.
         */

    }, {
        key: 'resolveInjection',
        value: function resolveInjection(givenInjection, buildConfigurations, modulesToExclude) {
            var moduleAliases = arguments.length <= 3 || arguments[3] === undefined ? {} : arguments[3];
            var knownExtensions = arguments.length <= 4 || arguments[4] === undefined ? ['', '.js', '.css', '.svg', '.html', 'json'] : arguments[4];
            var context = arguments.length <= 5 || arguments[5] === undefined ? './' : arguments[5];
            var referencePath = arguments.length <= 6 || arguments[6] === undefined ? '' : arguments[6];
            var pathsToIgnore = arguments.length <= 7 || arguments[7] === undefined ? ['.git'] : arguments[7];

            var injection = _clientnode2.default.extendObject(true, {}, givenInjection);
            var moduleFilePathsToExclude = Helper.determineModuleLocations(modulesToExclude, moduleAliases, knownExtensions, context, referencePath, pathsToIgnore).filePaths;
            var _arr2 = ['internal', 'external'];
            for (var _i2 = 0; _i2 < _arr2.length; _i2++) {
                var type = _arr2[_i2];
                /* eslint-disable curly */
                if (_typeof(injection[type]) === 'object') {
                    for (var chunkName in injection[type]) {
                        if (injection[type][chunkName] === '__auto__') {
                            injection[type][chunkName] = [];
                            var modules = Helper.getAutoChunk(buildConfigurations, moduleFilePathsToExclude, context);
                            for (var subChunkName in modules) {
                                if (modules.hasOwnProperty(subChunkName)) injection[type][chunkName].push(modules[subChunkName]);
                            } /*
                                  Reverse array to let javaScript files be the last
                                  ones to export them rather.
                              */
                            injection[type][chunkName].reverse();
                        }
                    }
                } else if (injection[type] === '__auto__')
                    /* eslint-enable curly */
                    injection[type] = Helper.getAutoChunk(buildConfigurations, moduleFilePathsToExclude, context);
            }return injection;
        }
        /**
         * Determines all module file paths.
         * @param buildConfigurations - Resolved build configuration.
         * @param moduleFilePathsToExclude - A list of modules file paths to
         * exclude (specified by path or id) or a mapping from chunk names to
         * module ids.
         * @param context - File path to use as starting point.
         * @returns All determined module file paths.
         */

    }, {
        key: 'getAutoChunk',
        value: function getAutoChunk(buildConfigurations, moduleFilePathsToExclude, context) {
            var result = {};
            var injectedBaseNames = {};
            var _iteratorNormalCompletion5 = true;
            var _didIteratorError5 = false;
            var _iteratorError5 = undefined;

            try {
                for (var _iterator5 = buildConfigurations[Symbol.iterator](), _step5; !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {
                    var buildConfiguration = _step5.value;

                    if (!injectedBaseNames[buildConfiguration.outputExtension]) injectedBaseNames[buildConfiguration.outputExtension] = [];
                    var _iteratorNormalCompletion6 = true;
                    var _didIteratorError6 = false;
                    var _iteratorError6 = undefined;

                    try {
                        for (var _iterator6 = buildConfiguration.filePaths[Symbol.iterator](), _step6; !(_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done); _iteratorNormalCompletion6 = true) {
                            var moduleFilePath = _step6.value;

                            if (!moduleFilePathsToExclude.includes(moduleFilePath)) {
                                var relativeModuleFilePath = _path2.default.relative(context, moduleFilePath);
                                var baseName = _path2.default.basename(relativeModuleFilePath, '.' + buildConfiguration.extension);
                                /*
                                    Ensure that each output type has only one source
                                    representation.
                                */
                                if (!injectedBaseNames[buildConfiguration.outputExtension].includes(baseName)) {
                                    /*
                                        Ensure that same basenames and different output
                                        types can be distinguished by their extension
                                        (JavaScript-Modules remains without extension since
                                        they will be handled first because the build
                                        configurations are expected to be sorted in this
                                        context).
                                    */
                                    if (result[baseName]) result[relativeModuleFilePath] = relativeModuleFilePath;else result[baseName] = relativeModuleFilePath;
                                    injectedBaseNames[buildConfiguration.outputExtension].push(baseName);
                                }
                            }
                        }
                    } catch (err) {
                        _didIteratorError6 = true;
                        _iteratorError6 = err;
                    } finally {
                        try {
                            if (!_iteratorNormalCompletion6 && _iterator6.return) {
                                _iterator6.return();
                            }
                        } finally {
                            if (_didIteratorError6) {
                                throw _iteratorError6;
                            }
                        }
                    }
                }
            } catch (err) {
                _didIteratorError5 = true;
                _iteratorError5 = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion5 && _iterator5.return) {
                        _iterator5.return();
                    }
                } finally {
                    if (_didIteratorError5) {
                        throw _iteratorError5;
                    }
                }
            }

            return result;
        }
        /**
         * Determines a concrete file path for given module id.
         * @param moduleID - Module id to determine.
         * @param moduleAliases - Mapping of aliases to take into account.
         * @param knownExtensions - List of known extensions.
         * @param context - File path to determine relative to.
         * @param referencePath - Path to resolve local modules relative to.
         * @param pathsToIgnore - Paths which marks location to ignore.
         * @param relativeModuleFilePaths - List of relative file path to search
         * for modules in.
         * @param packageEntryFileNames - List of package entry file names to
         * search for. The magic name "__package__" will search for an appreciate
         * entry in a "package.json" file.
         * @returns File path or given module id if determinations has failed or
         * wasn't necessary.
         */

    }, {
        key: 'determineModuleFilePath',
        value: function determineModuleFilePath(moduleID) {
            var moduleAliases = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];
            var knownExtensions = arguments.length <= 2 || arguments[2] === undefined ? ['', '.js', '.css', '.svg', '.html', 'json'] : arguments[2];
            var context = arguments.length <= 3 || arguments[3] === undefined ? './' : arguments[3];
            var referencePath = arguments.length <= 4 || arguments[4] === undefined ? '' : arguments[4];
            var pathsToIgnore = arguments.length <= 5 || arguments[5] === undefined ? ['.git'] : arguments[5];
            var relativeModuleFilePaths = arguments.length <= 6 || arguments[6] === undefined ? ['node_modules', '../'] : arguments[6];
            var packageEntryFileNames = arguments.length <= 7 || arguments[7] === undefined ? ['__package__', '', 'index', 'main'] : arguments[7];

            moduleID = Helper.applyAliases(Helper.stripLoader(moduleID), moduleAliases);
            if (!moduleID) return null;
            if (referencePath.startsWith('/')) referencePath = _path2.default.relative(context, referencePath);
            var _iteratorNormalCompletion7 = true;
            var _didIteratorError7 = false;
            var _iteratorError7 = undefined;

            try {
                for (var _iterator7 = [referencePath].concat(relativeModuleFilePaths)[Symbol.iterator](), _step7; !(_iteratorNormalCompletion7 = (_step7 = _iterator7.next()).done); _iteratorNormalCompletion7 = true) {
                    var moduleLocation = _step7.value;
                    var _iteratorNormalCompletion8 = true;
                    var _didIteratorError8 = false;
                    var _iteratorError8 = undefined;

                    try {
                        for (var _iterator8 = packageEntryFileNames[Symbol.iterator](), _step8; !(_iteratorNormalCompletion8 = (_step8 = _iterator8.next()).done); _iteratorNormalCompletion8 = true) {
                            var fileName = _step8.value;
                            var _iteratorNormalCompletion9 = true;
                            var _didIteratorError9 = false;
                            var _iteratorError9 = undefined;

                            try {
                                for (var _iterator9 = knownExtensions[Symbol.iterator](), _step9; !(_iteratorNormalCompletion9 = (_step9 = _iterator9.next()).done); _iteratorNormalCompletion9 = true) {
                                    var extension = _step9.value;

                                    var moduleFilePath = moduleID;
                                    if (!moduleFilePath.startsWith('/')) moduleFilePath = _path2.default.resolve(context, moduleLocation, moduleFilePath);
                                    if (fileName === '__package__') {
                                        try {
                                            if (fileSystem.statSync(moduleFilePath).isDirectory()) {
                                                var pathToPackageJSON = _path2.default.resolve(moduleFilePath, 'package.json');
                                                if (fileSystem.statSync(pathToPackageJSON).isFile()) {
                                                    var localConfiguration = JSON.parse(fileSystem.readFileSync(pathToPackageJSON, {
                                                        encoding: 'utf-8' }));
                                                    if (localConfiguration.main) fileName = localConfiguration.main;
                                                }
                                            }
                                        } catch (error) {}
                                        if (fileName === '__package__') continue;
                                    }
                                    moduleFilePath = _path2.default.resolve(moduleFilePath, fileName);
                                    moduleFilePath += extension;
                                    if (Helper.isFilePathInLocation(moduleFilePath, pathsToIgnore)) continue;
                                    if (Helper.isFileSync(moduleFilePath)) return moduleFilePath;
                                }
                            } catch (err) {
                                _didIteratorError9 = true;
                                _iteratorError9 = err;
                            } finally {
                                try {
                                    if (!_iteratorNormalCompletion9 && _iterator9.return) {
                                        _iterator9.return();
                                    }
                                } finally {
                                    if (_didIteratorError9) {
                                        throw _iteratorError9;
                                    }
                                }
                            }
                        }
                    } catch (err) {
                        _didIteratorError8 = true;
                        _iteratorError8 = err;
                    } finally {
                        try {
                            if (!_iteratorNormalCompletion8 && _iterator8.return) {
                                _iterator8.return();
                            }
                        } finally {
                            if (_didIteratorError8) {
                                throw _iteratorError8;
                            }
                        }
                    }
                }
            } catch (err) {
                _didIteratorError7 = true;
                _iteratorError7 = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion7 && _iterator7.return) {
                        _iterator7.return();
                    }
                } finally {
                    if (_didIteratorError7) {
                        throw _iteratorError7;
                    }
                }
            }

            return null;
        }
        // endregion
        /**
         * Determines a concrete file path for given module id.
         * @param moduleID - Module id to determine.
         * @param aliases - Mapping of aliases to take into account.
         * @returns The alias applied given module id.
         */

    }, {
        key: 'applyAliases',
        value: function applyAliases(moduleID, aliases) {
            for (var alias in aliases) {
                if (alias.endsWith('$')) {
                    if (moduleID === alias.substring(0, alias.length - 1)) moduleID = aliases[alias];
                } else moduleID = moduleID.replace(alias, aliases[alias]);
            }return moduleID;
        }
    }]);

    return Helper;
}();
// endregion
// region vim modline
// vim: set tabstop=4 shiftwidth=4 expandtab:
// vim: foldmethod=marker foldmarker=region,endregion:
// endregion


exports.default = Helper;
