// weboptimizer/pugLoader.compiled.js
#!/usr/bin/env node


// -*- coding: utf-8 -*-
'use strict';
/* !
    region header
    Copyright Torben Sickert (info["~at~"]torben.website) 16.12.2012

    License
    -------

    This library written by Torben Sickert stand under a creative commons naming
    3.0 unported license. see http://creativecommons.org/licenses/by/3.0/deed.de
    endregion
*/
// region imports

var _clientnode = require('clientnode');

var _clientnode2 = _interopRequireDefault(_clientnode);

var _fs = require('fs');

var fileSystem = _interopRequireWildcard(_fs);

var _loaderUtils = require('loader-utils');

var loaderUtils = _interopRequireWildcard(_loaderUtils);

var _pug = require('pug');

var pug = _interopRequireWildcard(_pug);

var _configurator = require('./configurator.compiled');

var _configurator2 = _interopRequireDefault(_configurator);

var _helper = require('./helper.compiled');

var _helper2 = _interopRequireDefault(_helper);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// NOTE: Only needed for debugging this file.
try {
    require('source-map-support/register');
} catch (error) {}

// endregion

// endregion
// region types
module.exports = function (source) {
    var _this = this;

    if (this.cacheable) this.cacheable();
    var query = _clientnode2.default.convertSubstringInPlainObject(_clientnode2.default.extendObject(true, {
        moduleAliases: [],
        knownExtensions: ['.pug', '.html', '.js', '.css'],
        context: './'
    }, this.options.pug || {}, loaderUtils.parseQuery(this.query)), /#%%%#/g, '!');
    var compile = function compile(template) {
        var options = arguments.length <= 1 || arguments[1] === undefined ? query.compiler : arguments[1];
        return function () {
            var locals = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

            options = _clientnode2.default.extendObject(true, {
                filename: template, doctype: 'html',
                compileDebug: _this.debug || false
            }, options);
            var templateFunction = void 0;
            if (options.isString) {
                delete options.isString;
                templateFunction = pug.compile(template, options);
            } else templateFunction = pug.compileFile(template, options);
            return templateFunction(_clientnode2.default.extendObject(true, {
                configuration: _configurator2.default, require: function require(request) {
                    var template = request.replace(/^(.+)\?[^?]+$/, '$1');
                    var queryMatch = request.match(/^[^?]+\?(.+)$/, '$1');
                    var nestedLocals = {};
                    if (queryMatch) {
                        var evaluationFunction = function evaluationFunction(request, template, source, compile, locals
                        // IgnoreTypeCheck
                        ) {
                            return new Function('request', 'template', 'source', 'compile', 'locals', 'return ' + queryMatch[1])(request, template, source, compile, locals);
                        };
                        nestedLocals = evaluationFunction(request, template, source, compile, locals);
                    }
                    var options = _clientnode2.default.extendObject(true, {
                        encoding: 'utf-8'
                    }, nestedLocals.options || {});
                    if (options.isString) return compile(template, options)(nestedLocals);
                    var templateFilePath = _helper2.default.determineModuleFilePath(template, query.moduleAliases, query.knownExtensions, query.context, _configurator2.default.path.source.asset.base, _configurator2.default.path.ignore);
                    if (templateFilePath) {
                        _this.addDependency(templateFilePath);
                        if (queryMatch || templateFilePath.endsWith('.pug')) return compile(templateFilePath, options)(nestedLocals);
                        return fileSystem.readFileSync(templateFilePath, options);
                    }
                    throw Error('Given template file "' + template + '" couldn\'t be resolved.');
                } }, locals));
        };
    };
    return compile(source, _clientnode2.default.extendObject(true, {
        isString: true,
        filename: loaderUtils.getRemainingRequest(this).replace(/^!/, '')
    }, query.compiler || {}))(query.locals || {});
};
// region vim modline
// vim: set tabstop=4 shiftwidth=4 expandtab:
// vim: foldmethod=marker foldmarker=region,endregion:
// endregion
