// weboptimizer/webpackConfigurator.compiled.js
#!/usr/bin/env node


// -*- coding: utf-8 -*-
'use strict';
/* !
    region header
    Copyright Torben Sickert (info["~at~"]torben.website) 16.12.2012

    License
    -------

    This library written by Torben Sickert stand under a creative commons naming
    3.0 unported license. see http://creativecommons.org/licenses/by/3.0/deed.de
    endregion
*/
// region imports

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _clientnode = require('clientnode');

var _clientnode2 = _interopRequireDefault(_clientnode);

var _jsdom = require('jsdom');

var dom = _interopRequireWildcard(_jsdom);

var _fs = require('fs');

var fileSystem = _interopRequireWildcard(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _postcssCssnext = require('postcss-cssnext');

var _postcssCssnext2 = _interopRequireDefault(_postcssCssnext);

var _postcssFontpath = require('postcss-fontpath');

var _postcssFontpath2 = _interopRequireDefault(_postcssFontpath);

var _postcssImport = require('postcss-import');

var _postcssImport2 = _interopRequireDefault(_postcssImport);

var _postcssSprites = require('postcss-sprites');

var _postcssSprites2 = _interopRequireDefault(_postcssSprites);

var _postcssUrl = require('postcss-url');

var _postcssUrl2 = _interopRequireDefault(_postcssUrl);

var _util = require('util');

var _util2 = _interopRequireDefault(_util);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _webpackSources = require('webpack-sources');

var _configurator = require('./configurator.compiled');

var _configurator2 = _interopRequireDefault(_configurator);

var _helper = require('./helper.compiled');

var _helper2 = _interopRequireDefault(_helper);

var _htmlLoader = require('html-loader');

var _htmlLoader2 = _interopRequireDefault(_htmlLoader);

var _loaderUtils = require('loader-utils');

var _loaderUtils2 = _interopRequireDefault(_loaderUtils);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// NOTE: Only needed for debugging this file.
try {
    require('source-map-support/register');
} catch (error) {}
var plugins = require('webpack-load-plugins')();


plugins.HTML = plugins.html;
plugins.ExtractText = plugins.extractText;
plugins.AddAssetHTMLPlugin = require('add-asset-html-webpack-plugin');
plugins.OpenBrowser = plugins.openBrowser;
plugins.Favicon = require('favicons-webpack-plugin');
plugins.Imagemin = require('imagemin-webpack-plugin').default;
plugins.Offline = require('offline-plugin');

// / region monkey patches
// Monkey-Patch html loader to retrieve html loader options since the
// "webpack-html-plugin" doesn't preserve the original loader interface.

require.cache[require.resolve('html-loader')].exports = function () {
    _clientnode2.default.extendObject(true, this.options, module, this.options);
    return _htmlLoader2.default.apply(this, arguments);
};
// Monkey-Patch loader-utils to define which url is a local request.

var loaderUtilsIsUrlRequestBackup = _loaderUtils2.default.isUrlRequest;
require.cache[require.resolve('loader-utils')].exports.isUrlRequest = function (url) {
    if (url.match(/^[a-z]+:.+/)) return false;
    return loaderUtilsIsUrlRequestBackup.apply(_loaderUtils2.default, arguments);
};
// / endregion
// endregion
// region initialisation
// / region determine library name
var libraryName = void 0;
if ('libraryName' in _configurator2.default && _configurator2.default.libraryName) libraryName = _configurator2.default.libraryName;else if (Object.keys(_configurator2.default.injection.internal.normalized).length > 1) libraryName = '[name]';else {
    libraryName = _configurator2.default.name;
    if (_configurator2.default.exportFormat.self === 'var') libraryName = _clientnode2.default.stringConvertToValidVariableName(libraryName);
}
// / endregion
// / region plugins
var pluginInstances = [new _webpack2.default.optimize.OccurrenceOrderPlugin(true)];
// // region define modules to ignore
var _iteratorNormalCompletion = true;
var _didIteratorError = false;
var _iteratorError = undefined;

try {
    for (var _iterator = _configurator2.default.injection.ignorePattern[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var ignorePattern = _step.value;

        pluginInstances.push(new _webpack2.default.IgnorePlugin(new RegExp(ignorePattern)));
    } // // endregion
    // // region generate html file
} catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
} finally {
    try {
        if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
        }
    } finally {
        if (_didIteratorError) {
            throw _iteratorError;
        }
    }
}

var htmlAvailable = false;
if (_configurator2.default.givenCommandLineArguments[2] !== 'buildDLL') {
    var _iteratorNormalCompletion2 = true;
    var _didIteratorError2 = false;
    var _iteratorError2 = undefined;

    try {
        for (var _iterator2 = _configurator2.default.files.html[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
            var _htmlConfiguration = _step2.value;

            if (_helper2.default.isFileSync(_htmlConfiguration.template.substring(_htmlConfiguration.template.lastIndexOf('!') + 1))) {
                if (_htmlConfiguration.template === _configurator2.default.files.defaultHTML.template) _htmlConfiguration.template = _htmlConfiguration.template.substring(_htmlConfiguration.template.lastIndexOf('!') + 1);
                pluginInstances.push(new plugins.HTML(_htmlConfiguration));
                htmlAvailable = true;
            }
        }
    } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
    } finally {
        try {
            if (!_iteratorNormalCompletion2 && _iterator2.return) {
                _iterator2.return();
            }
        } finally {
            if (_didIteratorError2) {
                throw _iteratorError2;
            }
        }
    }
} // // endregion
// // region generate favicons
if (htmlAvailable && _configurator2.default.favicon && _helper2.default.isFileSync(_configurator2.default.favicon.logo)) pluginInstances.push(new plugins.Favicon(_configurator2.default.favicon));
// // endregion
// // region provide offline functionality
if (htmlAvailable && _configurator2.default.offline) {
    if (!['serve', 'testInBrowser'].includes(_configurator2.default.givenCommandLineArguments[2])) {
        if (_configurator2.default.inPlace.cascadingStyleSheet) _configurator2.default.offline.excludes.push(_path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.path.target.asset.cascadingStyleSheet) + ('*.css?' + _configurator2.default.hashAlgorithm + '=*'));
        if (_configurator2.default.inPlace.javaScript) _configurator2.default.offline.excludes.push(_path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.path.target.asset.javaScript) + ('*.js?' + _configurator2.default.hashAlgorithm + '=*'));
    }
    pluginInstances.push(new plugins.Offline(_configurator2.default.offline));
}
// // endregion
// // region opens browser automatically
if (_configurator2.default.development.openBrowser && htmlAvailable && ['serve', 'testInBrowser'].includes(_configurator2.default.givenCommandLineArguments[2])) pluginInstances.push(new plugins.OpenBrowser(_configurator2.default.development.openBrowser));
// // endregion
// // region provide build environment
if (_configurator2.default.buildDefinition) pluginInstances.push(new _webpack2.default.DefinePlugin(_configurator2.default.buildDefinition));
if (_configurator2.default.module.provide) pluginInstances.push(new _webpack2.default.ProvidePlugin(_configurator2.default.module.provide));
// // endregion
// // region modules/assets
// /// region perform javaScript minification/optimisation
if (_configurator2.default.module.optimizer.uglifyJS) pluginInstances.push(new _webpack2.default.optimize.UglifyJsPlugin(_configurator2.default.module.optimizer.uglifyJS));
// /// endregion
// /// region apply module pattern
pluginInstances.push({ apply: function apply(compiler) {
        compiler.plugin('emit', function (compilation, callback) {
            for (var request in compilation.assets) {
                if (compilation.assets.hasOwnProperty(request)) {
                    var filePath = request.replace(/\?[^?]+$/, '');
                    var type = _helper2.default.determineAssetType(filePath, _configurator2.default.build, _configurator2.default.path);
                    if (type && _configurator2.default.assetPattern[type] && !new RegExp(_configurator2.default.assetPattern[type].excludeFilePathRegularExpression).test(filePath)) {
                        var source = compilation.assets[request].source();
                        if (typeof source === 'string') compilation.assets[request] = new _webpackSources.RawSource(_configurator2.default.assetPattern[type].pattern.replace(/\{1\}/g, source.replace(/\$/g, '$$$')));
                    }
                }
            }callback();
        });
    } });
// /// endregion
// /// region in-place configured assets in the main html file
if (htmlAvailable && !['serve', 'testInBrowser'].includes(_configurator2.default.givenCommandLineArguments[2])) pluginInstances.push({ apply: function apply(compiler) {
        compiler.plugin('emit', function (compilation, callback) {
            if (_configurator2.default.files.html[0].filename in compilation.assets && (_configurator2.default.inPlace.cascadingStyleSheet || _configurator2.default.inPlace.javaScript)) dom.env(compilation.assets[_configurator2.default.files.html[0].filename].source(), function (error, window) {
                if (_configurator2.default.inPlace.cascadingStyleSheet) {
                    var urlPrefix = _path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.files.compose.cascadingStyleSheet.replace('[contenthash]', ''));
                    var domNode = window.document.querySelector('link[href^="' + urlPrefix + '"]');
                    if (domNode) {
                        var asset = void 0;
                        for (asset in compilation.assets) {
                            if (asset.startsWith(urlPrefix)) break;
                        }var inPlaceDomNode = window.document.createElement('style');
                        inPlaceDomNode.textContent = compilation.assets[asset].source();
                        domNode.parentNode.insertBefore(inPlaceDomNode, domNode);
                        domNode.parentNode.removeChild(domNode);
                        /*
                            NOTE: This doesn't prevent webpack from
                            creating this file if present in another chunk
                            so removing it (and a potential source map
                            file) later in the "done" hook.
                        */
                        delete compilation.assets[asset];
                    } else console.warn('No referenced cascading style sheet file in' + ' resulting markup found with ' + ('selector: link[href^="' + urlPrefix + '"]'));
                }
                if (_configurator2.default.inPlace.javaScript) {
                    var _urlPrefix = _path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.files.compose.javaScript.replace('[hash]', ''));
                    var _domNode = window.document.querySelector('script[src^="' + _urlPrefix + '"]');
                    if (_domNode) {
                        var _asset = void 0;
                        for (_asset in compilation.assets) {
                            if (_asset.startsWith(_urlPrefix)) break;
                        }_domNode.textContent = compilation.assets[_asset].source();
                        _domNode.removeAttribute('src');
                        /*
                            NOTE: This doesn't prevent webpack from
                            creating this file if present in another chunk
                            so removing it (and a potential source map
                            file) later in the "done" hook.
                        */
                        delete compilation.assets[_asset];
                    } else console.warn('No referenced javaScript file in resulting ' + 'markup found with selector: ' + ('script[src^="' + _urlPrefix + '"]'));
                }
                compilation.assets[_configurator2.default.files.html[0].filename] = new _webpackSources.RawSource(compilation.assets[_configurator2.default.files.html[0].filename].source().replace(/^(\s*<!doctype[^>]+?>\s*)[\s\S]*$/i, '$1') + window.document.documentElement.outerHTML);
                callback();
            });else callback();
        });
        compiler.plugin('after-emit', function (compilation, callback) {
            if (_configurator2.default.files.html[0].filename in compilation.assets) {
                if (_configurator2.default.inPlace.cascadingStyleSheet) {
                    var assetFilePath = _helper2.default.stripLoader(_configurator2.default.files.compose.cascadingStyleSheet);
                    if (_helper2.default.isFileSync(assetFilePath)) fileSystem.unlinkSync(assetFilePath);
                }
                if (_configurator2.default.inPlace.javaScript) {
                    var assetFilePathTemplate = _helper2.default.stripLoader(_configurator2.default.files.compose.javaScript);
                    for (var chunkName in _configurator2.default.injection.internal.normalized) {
                        if (_configurator2.default.injection.internal.normalized.hasOwnProperty(chunkName)) {
                            var _assetFilePath = _helper2.default.renderFilePathTemplate(assetFilePathTemplate, { '[name]': chunkName });
                            if (_helper2.default.isFileSync(_assetFilePath)) fileSystem.unlinkSync(_assetFilePath);
                        }
                    }
                }
                var _arr = ['javaScript', 'cascadingStyleSheet'];
                for (var _i = 0; _i < _arr.length; _i++) {
                    var type = _arr[_i];
                    if (fileSystem.readdirSync(_configurator2.default.path.target.asset[type]).length === 0) fileSystem.rmdirSync(_configurator2.default.path.target.asset[type]);
                }
            }
            callback();
        });
    } });
// /// endregion
// /// region remove chunks if a corresponding dll package exists
if (_configurator2.default.givenCommandLineArguments[2] !== 'buildDLL') for (var chunkName in _configurator2.default.injection.internal.normalized) {
    if (_configurator2.default.injection.internal.normalized.hasOwnProperty(chunkName)) {
        var manifestFilePath = _configurator2.default.path.target.base + '/' + chunkName + '.' + 'dll-manifest.json';
        if (_configurator2.default.dllManifestFilePaths.includes(manifestFilePath)) {
            delete _configurator2.default.injection.internal.normalized[chunkName];
            var filePath = _helper2.default.renderFilePathTemplate(_helper2.default.stripLoader(_configurator2.default.files.compose.javaScript), { '[name]': chunkName });
            pluginInstances.push(new plugins.AddAssetHTMLPlugin({
                filepath: filePath,
                hash: true,
                includeSourcemap: _helper2.default.isFileSync(filePath + '.map')
            }));
            pluginInstances.push(new _webpack2.default.DllReferencePlugin({
                context: _configurator2.default.path.context, manifest: require(manifestFilePath) }));
        }
    }
} // /// endregion
// /// region generate common chunks
if (_configurator2.default.givenCommandLineArguments[2] !== 'buildDLL') {
    var _iteratorNormalCompletion3 = true;
    var _didIteratorError3 = false;
    var _iteratorError3 = undefined;

    try {
        for (var _iterator3 = _configurator2.default.injection.commonChunkIDs[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
            var _chunkName = _step3.value;

            if (_configurator2.default.injection.internal.normalized.hasOwnProperty(_chunkName)) pluginInstances.push(new _webpack2.default.optimize.CommonsChunkPlugin({
                async: false,
                children: false,
                filename: _path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.files.compose.javaScript),
                minChunks: Infinity,
                name: _chunkName,
                minSize: 0
            }));
        }
    } catch (err) {
        _didIteratorError3 = true;
        _iteratorError3 = err;
    } finally {
        try {
            if (!_iteratorNormalCompletion3 && _iterator3.return) {
                _iterator3.return();
            }
        } finally {
            if (_didIteratorError3) {
                throw _iteratorError3;
            }
        }
    }
} // /// endregion
// /// region mark empty javaScript modules as dummy
if (!_configurator2.default.needed.javaScript) _configurator2.default.files.compose.javaScript = _path2.default.resolve(_configurator2.default.path.target.asset.javaScript, '.__dummy__.compiled.js');
// /// endregion
// /// region extract cascading style sheets
pluginInstances.push(new plugins.ExtractText(_configurator2.default.files.compose.cascadingStyleSheet ? _path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.files.compose.cascadingStyleSheet) : _configurator2.default.path.target.base, { allChunks: true, disable: !_configurator2.default.files.compose.cascadingStyleSheet }));
// /// endregion
// /// region performs implicit external logic
if (_configurator2.default.injection.external === '__implicit__')
    /*
        We only want to process modules from local context in library mode,
        since a concrete project using this library should combine all assets
        (and deduplicate them) for optimal bundling results. NOTE: Only native
        javaScript and json modules will be marked as external dependency.
    */
    _configurator2.default.injection.external = function (context, request, callback) {
        request = request.replace(/^!+/, '');
        if (request.startsWith('/')) request = _path2.default.relative(_configurator2.default.path.context, request);
        var _iteratorNormalCompletion4 = true;
        var _didIteratorError4 = false;
        var _iteratorError4 = undefined;

        try {
            for (var _iterator4 = _configurator2.default.module.directories.concat(_configurator2.default.loader.directories)[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
                var _filePath = _step4.value;

                if (request.startsWith(_filePath)) {
                    request = request.substring(_filePath.length);
                    if (request.startsWith('/')) request = request.substring(1);
                    break;
                }
            }
        } catch (err) {
            _didIteratorError4 = true;
            _iteratorError4 = err;
        } finally {
            try {
                if (!_iteratorNormalCompletion4 && _iterator4.return) {
                    _iterator4.return();
                }
            } finally {
                if (_didIteratorError4) {
                    throw _iteratorError4;
                }
            }
        }

        var resolvedRequest = _helper2.default.determineExternalRequest(request, _configurator2.default.path.context, context, _configurator2.default.injection.internal.normalized, _configurator2.default.path.ignore.concat(_configurator2.default.module.directories, _configurator2.default.loader.directories).map(function (filePath) {
            return _path2.default.resolve(_configurator2.default.path.context, filePath);
        }).filter(function (filePath) {
            return !_configurator2.default.path.context.startsWith(filePath);
        }), _configurator2.default.module.aliases, _configurator2.default.knownExtensions, _configurator2.default.path.source.asset.base, _configurator2.default.path.ignore, _configurator2.default.injection.implicitExternalIncludePattern, _configurator2.default.injection.implicitExternalExcludePattern, _configurator2.default.inPlace.externalLibrary.normal, _configurator2.default.inPlace.externalLibrary.dynamic);
        if (resolvedRequest) {
            if (['var', 'umd'].includes(_configurator2.default.exportFormat.external) && request in _configurator2.default.injection.externalAliases) resolvedRequest = _configurator2.default.injection.externalAliases[request];
            if (_configurator2.default.exportFormat.external === 'var') resolvedRequest = _clientnode2.default.stringConvertToValidVariableName(resolvedRequest, '0-9a-zA-Z_$\\.');
            return callback(null, resolvedRequest, _configurator2.default.exportFormat.external);
        }
        return callback();
    };
// /// endregion
// /// region build dll packages
if (_configurator2.default.givenCommandLineArguments[2] === 'buildDLL') {
    var dllChunkIDExists = false;
    for (var _chunkName2 in _configurator2.default.injection.internal.normalized) {
        if (_configurator2.default.injection.internal.normalized.hasOwnProperty(_chunkName2)) if (_configurator2.default.injection.dllChunkIDs.includes(_chunkName2)) dllChunkIDExists = true;else delete _configurator2.default.injection.internal.normalized[_chunkName2];
    }if (dllChunkIDExists) {
        libraryName = '[name]DLLPackage';
        pluginInstances.push(new _webpack2.default.DllPlugin({
            path: _configurator2.default.path.target.base + '/[name].dll-manifest.json',
            name: libraryName
        }));
    } else console.warn('No dll chunk id found.');
}
// /// endregion
// // endregion
// // region apply final dom/javaScript modifications/fixes
pluginInstances.push({ apply: function apply(compiler) {
        compiler.plugin('emit', function (compilation, callback) {
            var promises = [];
            /*
                NOTE: Removing symbols after a "&" in hash string is necessary to
                match the generated request strings in offline plugin.
            */
            var _iteratorNormalCompletion5 = true;
            var _didIteratorError5 = false;
            var _iteratorError5 = undefined;

            try {
                var _loop = function _loop() {
                    var htmlConfiguration = _step5.value;

                    if (htmlConfiguration.filename in compilation.assets) promises.push(new Promise(function (resolve, reject) {
                        return dom.env(compilation.assets[htmlConfiguration.filename].source(), function (error, window) {
                            if (error) return reject(error);
                            var linkables = {
                                script: 'src', link: 'href' };
                            for (var tagName in linkables) {
                                if (linkables.hasOwnProperty(tagName)) {
                                    var _iteratorNormalCompletion6 = true;
                                    var _didIteratorError6 = false;
                                    var _iteratorError6 = undefined;

                                    try {
                                        for (var _iterator6 = window.document.querySelectorAll(tagName + '[' + linkables[tagName] + '*="?' + (_configurator2.default.hashAlgorithm + '="]'))[Symbol.iterator](), _step6; !(_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done); _iteratorNormalCompletion6 = true) {
                                            var _domNode2 = _step6.value;

                                            _domNode2.setAttribute(linkables[tagName], _domNode2.getAttribute(linkables[tagName]).replace(new RegExp('(\\?' + _configurator2.default.hashAlgorithm + '=' + '[^&]+).*$'), '$1'));
                                        }
                                    } catch (err) {
                                        _didIteratorError6 = true;
                                        _iteratorError6 = err;
                                    } finally {
                                        try {
                                            if (!_iteratorNormalCompletion6 && _iterator6.return) {
                                                _iterator6.return();
                                            }
                                        } finally {
                                            if (_didIteratorError6) {
                                                throw _iteratorError6;
                                            }
                                        }
                                    }
                                }
                            }compilation.assets[htmlConfiguration.filename] = new _webpackSources.RawSource(compilation.assets[htmlConfiguration.filename].source().replace(/^(\s*<!doctype[^>]+?>\s*)[\s\S]*$/i, '$1') + window.document.documentElement.outerHTML);
                            return resolve(compilation.assets[htmlConfiguration.filename]);
                        });
                    }));
                };

                for (var _iterator5 = _configurator2.default.files.html[Symbol.iterator](), _step5; !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {
                    _loop();
                }
            } catch (err) {
                _didIteratorError5 = true;
                _iteratorError5 = err;
            } finally {
                try {
                    if (!_iteratorNormalCompletion5 && _iterator5.return) {
                        _iterator5.return();
                    }
                } finally {
                    if (_didIteratorError5) {
                        throw _iteratorError5;
                    }
                }
            }

            if (!_configurator2.default.exportFormat.external.startsWith('umd')) {
                Promise.all(promises).then(function () {
                    return callback();
                });
                return;
            }
            var bundleName = typeof libraryName === 'string' ? libraryName : libraryName[0];
            /*
                NOTE: The umd module export doesn't handle cases where the package
                name doesn't match exported library name. This post processing
                fixes this issue.
            */
            for (var assetRequest in compilation.assets) {
                if (assetRequest.replace(/([^?]+)\?.*$/, '$1').endsWith(_configurator2.default.build.javaScript.outputExtension)) {
                    var source = compilation.assets[assetRequest].source();
                    if (typeof source === 'string') {
                        for (var replacement in _configurator2.default.injection.externalAliases) {
                            if (_configurator2.default.injection.externalAliases.hasOwnProperty(replacement)) source = source.replace(new RegExp('(require\\()"' + _clientnode2.default.stringConvertToValidRegularExpression(_configurator2.default.injection.externalAliases[replacement]) + '"(\\))', 'g'), '$1\'' + replacement + '\'$2').replace(new RegExp('(define\\("' + _clientnode2.default.stringConvertToValidRegularExpression(bundleName) + '", \\[.*)"' + _clientnode2.default.stringConvertToValidRegularExpression(_configurator2.default.injection.externalAliases[replacement]) + '"(.*\\], factory\\);)'), '$1\'' + replacement + '\'$2');
                        }source = source.replace(new RegExp('(root\\[)"' + _clientnode2.default.stringConvertToValidRegularExpression(bundleName) + '"(\\] = )'), "$1'" + _clientnode2.default.stringConvertToValidVariableName(bundleName) + "'$2");
                        compilation.assets[assetRequest] = new _webpackSources.RawSource(source);
                    }
                }
            }Promise.all(promises).then(function () {
                return callback();
            });
        });
    } });
// // endregion
// // region add automatic image compression
// NOTE: This plugin should be loaded at last to ensure that all emitted images
// ran through.
pluginInstances.push(new plugins.Imagemin(_configurator2.default.module.optimizer.image.content));
// // endregion
// / endregion
// / region loader
var imageLoader = 'url?' + _clientnode2.default.convertCircularObjectToJSON(_configurator2.default.module.optimizer.image.file);
var loader = {
    preprocessor: {
        cascadingStyleSheet: 'postcss' + _configurator2.default.module.preprocessor.cascadingStyleSheet,
        javaScript: 'babel?' + _clientnode2.default.convertCircularObjectToJSON(_configurator2.default.module.preprocessor.babel),
        pug: 'pug?' + _clientnode2.default.convertCircularObjectToJSON(_configurator2.default.module.preprocessor.pug)
    },
    html: 'html?' + _clientnode2.default.convertCircularObjectToJSON(_configurator2.default.module.html),
    cascadingStyleSheet: 'css' + _configurator2.default.module.cascadingStyleSheet,
    style: 'style?' + _clientnode2.default.convertCircularObjectToJSON(_configurator2.default.module.style),
    postprocessor: {
        image: imageLoader,
        font: {
            eot: 'url?' + _clientnode2.default.convertCircularObjectToJSON(_configurator2.default.module.optimizer.font.eot),
            woff: 'url?' + _clientnode2.default.convertCircularObjectToJSON(_configurator2.default.module.optimizer.font.woff),
            ttf: 'url?' + _clientnode2.default.convertCircularObjectToJSON(_configurator2.default.module.optimizer.font.ttf),
            svg: 'url?' + _clientnode2.default.convertCircularObjectToJSON(_configurator2.default.module.optimizer.font.svg)
        },
        data: 'url?' + _clientnode2.default.convertCircularObjectToJSON(_configurator2.default.module.optimizer.data)
    }
};
// / endregion
// endregion
// region configuration
var webpackConfiguration = {
    context: _configurator2.default.path.context,
    debug: _configurator2.default.debug,
    devtool: _configurator2.default.development.tool,
    devServer: _configurator2.default.development.server,
    // region input
    entry: _configurator2.default.injection.internal.normalized,
    externals: _configurator2.default.injection.external,
    resolveLoader: {
        alias: _configurator2.default.loader.aliases,
        extensions: _configurator2.default.loader.extensions,
        modulesDirectories: _configurator2.default.loader.directories
    },
    resolve: {
        alias: _configurator2.default.module.aliases,
        extensions: _configurator2.default.knownExtensions,
        root: [_configurator2.default.path.source.asset.base],
        modulesDirectories: _configurator2.default.module.directories
    },
    // endregion
    // region output
    output: {
        filename: _path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.files.compose.javaScript),
        hashFunction: _configurator2.default.hashAlgorithm,
        library: libraryName,
        libraryTarget: _configurator2.default.givenCommandLineArguments[2] === 'buildDLL' ? 'var' : _configurator2.default.exportFormat.self,
        path: _configurator2.default.path.target.base,
        publicPath: _configurator2.default.path.target.public,
        pathinfo: _configurator2.default.debug,
        umdNamedDefine: true
    },
    target: _configurator2.default.targetTechnology,
    // endregion
    module: {
        noParse: _configurator2.default.module.skipParseRegularExpression,
        preLoaders: [
        // Convert to native web types.
        // region script
        {
            test: /\.js(?:\?.*)?$/,
            loader: loader.preprocessor.javaScript,
            include: _helper2.default.normalizePaths([_configurator2.default.path.source.asset.javaScript].concat(_configurator2.default.module.locations.directoryPaths)),
            exclude: function exclude(filePath) {
                filePath = _helper2.default.stripLoader(filePath);
                return _helper2.default.isFilePathInLocation(filePath, _configurator2.default.path.ignore.concat(_configurator2.default.module.directories, _configurator2.default.loader.directories).map(function (filePath) {
                    return _path2.default.resolve(_configurator2.default.path.context, filePath);
                }).filter(function (filePath) {
                    return !_configurator2.default.path.context.startsWith(filePath);
                }));
            }
        },
        // endregion
        // region html (templates)
        // NOTE: This is only for the main entry template.
        {
            test: new RegExp('^' + _clientnode2.default.stringConvertToValidRegularExpression(_helper2.default.stripLoader(_configurator2.default.files.defaultHTML.template)) + '(?:\\?.*)?$'),
            loader: _configurator2.default.files.defaultHTML.template.substring(0, _configurator2.default.files.defaultHTML.template.lastIndexOf('!'))
        }, {
            test: /\.pug(?:\?.*)?$/,
            loader: 'file?name=' + _path2.default.relative(_configurator2.default.path.target.asset.base, _configurator2.default.path.target.asset.template) + ('[name].html?' + _configurator2.default.hashAlgorithm + '=[hash]!') + ('extract!' + loader.html + '!' + loader.preprocessor.pug),
            include: _configurator2.default.path.source.asset.template,
            exclude: _helper2.default.normalizePaths(_configurator2.default.files.html.concat(_configurator2.default.files.defaultHTML).map(function (htmlConfiguration) {
                return _helper2.default.stripLoader(htmlConfiguration.template);
            }))
        }
        // endregion
        ],
        loaders: [
        // Loads dependencies.
        // region style
        {
            test: /\.css(?:\?.*)?$/,
            loader: plugins.ExtractText.extract(loader.style, loader.cascadingStyleSheet + '!' + loader.preprocessor.cascadingStyleSheet),
            include: _helper2.default.normalizePaths([_configurator2.default.path.source.asset.cascadingStyleSheet].concat(_configurator2.default.module.locations.directoryPaths)),
            exclude: function exclude(filePath) {
                filePath = _helper2.default.stripLoader(filePath);
                return _helper2.default.isFilePathInLocation(filePath, _configurator2.default.path.ignore.concat(_configurator2.default.module.directories, _configurator2.default.loader.directories).map(function (filePath) {
                    return _path2.default.resolve(_configurator2.default.path.context, filePath);
                }).filter(function (filePath) {
                    return !_configurator2.default.path.context.startsWith(filePath);
                }));
            }
        },
        // endregion
        // region html (templates)
        {
            test: /\.html(?:\?.*)?$/,
            loader: 'file?name=' + _path2.default.relative(_configurator2.default.path.target.base, _configurator2.default.path.target.asset.template) + ('[name].[ext]?' + _configurator2.default.hashAlgorithm + '=[hash]!') + ('extract!' + loader.html),
            include: _configurator2.default.path.source.asset.template,
            exclude: _helper2.default.normalizePaths(_configurator2.default.files.html.map(function (htmlConfiguration) {
                return _helper2.default.stripLoader(htmlConfiguration.template);
            }))
        }
        // endregion
        ],
        postLoaders: [
        // Optimize loaded assets.
        // region font
        {
            test: /\.eot(?:\?.*)?$/,
            loader: loader.postprocessor.font.eot
        }, {
            test: /\.woff2?(?:\?.*)?$/,
            loader: loader.postprocessor.font.woff
        }, {
            test: /\.ttf(?:\?.*)?$/,
            loader: loader.postprocessor.font.ttf
        }, {
            test: /\.svg(?:\?.*)?$/,
            loader: loader.postprocessor.font.svg
        },
        // endregion
        // region image
        {
            test: /\.(?:png|jpg|ico|gif)(?:\?.*)?$/,
            loader: loader.postprocessor.image
        },
        // endregion
        // region data
        {
            test: /.+/,
            loader: loader.postprocessor.data,
            include: _configurator2.default.path.source.asset.data,
            exclude: function exclude(filePath) {
                return _configurator2.default.knownExtensions.includes(_path2.default.extname(_helper2.default.stripLoader(filePath)));
            }
        }
        // endregion
        ]
    },
    postcss: function postcss() {
        return [(0, _postcssImport2.default)({
            addDependencyTo: _webpack2.default,
            root: _configurator2.default.path.context
        }),
        /*
            NOTE: Checking path doesn't work if fonts are referenced in
            libraries provided in another location than the project itself like
            the node_modules folder.
        */
        (0, _postcssCssnext2.default)({ browsers: '> 0%' }), (0, _postcssFontpath2.default)({ checkPath: false }), (0, _postcssUrl2.default)({ filter: '', maxSize: 0 }), (0, _postcssSprites2.default)({
            filterBy: function filterBy() {
                return new Promise(function (resolve, reject) {
                    return (_configurator2.default.files.compose.image ? resolve : reject)();
                });
            },
            hooks: { onSaveSpritesheet: function onSaveSpritesheet(image) {
                    return _path2.default.join(image.spritePath, _path2.default.relative(_configurator2.default.path.target.asset.image, _configurator2.default.files.compose.image));
                } },
            stylesheetPath: _configurator2.default.path.source.asset.cascadingStyleSheet,
            spritePath: _configurator2.default.path.source.asset.image
        })];
    },
    html: _configurator2.default.module.optimizer.htmlMinifier,
    // Let the "html-loader" access full html minifier processing
    // configuration.
    pug: _configurator2.default.module.preprocessor.pug,
    plugins: pluginInstances
};
if (_configurator2.default.debug) console.log('Using webpack configuration:', _util2.default.inspect(webpackConfiguration, { depth: null }));
// endregion
exports.default = webpackConfiguration;
// region vim modline
// vim: set tabstop=4 shiftwidth=4 expandtab:
// vim: foldmethod=marker foldmarker=region,endregion:
// endregion
